Issue no 1195

jcogs-design opened the pull request:
<!--
ExpressionEngine uses semantic versioning.
- Small additive changes should target the next minor branch (release/next-minor if a numbered branch does not yet exist)
-->

<!-- What's in this pull request? -->
## Overview

A change to code for `template_post_parse` hook in Template class.
Current code fires after on sub or layout templates are parsed (i.e. during parsing process if multiple templates involved) but provides no mechanism to access the partial template output.  The parsed template is only made available via the hook after the 'final' template parse.

The updated code makes two changes:

- first it adds an extra parameter to the `template_post_parse` hook (`$currentTemplateInfo = array()`) that contains the name of the most recently completed template parse and template group it is from;
- second it modifies the information provided via the $template parameter defined in the hook: when hook fires on a partial template parse it returns the current working version of template

<!-- If this pull request resolves a project issue, provide a link: -->
Resolves [#1195](https://github.com/ExpressionEngine/ExpressionEngine/issues/1195).

## Nature of This Change

<!-- Check all that apply: -->

- [ ] 🐛 Fixes a bug
- [X] 🚀 Implements a new feature
- [X] 🛁 Refactors existing code
- [ ] 💅 Fixes coding style
- [ ] ✅ Adds tests
- [ ] 👽 Adds new dependency
- [ ] 🔥 Removes unused files / code
- [ ] 🔒 Improves security <!-- if your fix would EXPOSE a current security flaw, do not submit a pull request. Instead report a security bug at https://docs.expressionengine.com/latest/bugs_and_security_reports -->

## Is this backwards compatible?

- [X] Yes
- [ ] No

## Documentation
<!-- Required for new features -->
User Guide Pull Request: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pulls/217

<!-- Don't forget to add a single line changelog for your change to the appropriate file!

- changelogs/patch.rst (x.x.X)
- changelogs/minor.rst (x.X.x)
- changelogs/major.rst (X.x.x)

If you have not already, please sign the Contributor License Agreement: https://www.clahub.com/agreements/ExpressionEngine/ExpressionEngine

Thank you for contributing to ExpressionEngine! -->

jcogs-design committed:
Modify template_post_parse hook per issue #1195
Write current template name / group to array variable
Add conditional to return WIP template when not final

jcogs-design committed:
Merge branch '6.dev' of https://github.com/jcogs-design/ExpressionEngine into 6.dev

jcogs-design committed:
Remove redundant code
Original idea was to use count($this->templates_loaded)
to provide access to current templates, but actually they
appear via $template / $template_group anyhow...

jcogs-design committed:
Issue #1195
Update patch changelog

intoeetive commented on system/ee/legacy/libraries/Template.php:
```diff
@@ -1889,7 +1914,7 @@ public function fetch_cache_file($cfile, $cache_type = 'tag', $args = array())
 
         // Bail out if this tag/template isn't set to cache
-        if (ee('LivePreview')->hasEntryData() or ! isset($args['cache']) or $args['cache'] != 'yes') {
+        if (! isset($args['cache']) or $args['cache'] != 'yes') {
```
what is the reason check for live preview was taken out?

jcogs-design replied:
Was not an intended change - perhaps it is a sync issue and I was working from an old version of code?

jcogs-design replied:
On checking my copy of 6.dev the code is indeed missing - I seem to have got an old copy of the file somehow.  Will fix my local copy...

intoeetive commented on system/ee/legacy/libraries/Template.php:
```diff
@@ -232,15 +232,40 @@ public function fetch_and_parse($template_group = '', $template = '', $is_embed
+            // Add a conditional to adjust what is returned by function depending on what kind of template we are dealing with
+            if ($this->final_template)
+            {
+                $this->final_template = ee()->extensions->call(
+                    'template_post_parse',
+                    $this->final_template,
+                    ($is_embed || $is_layout), // $is_partial
+                    $site_id,
+                    $currentTemplateInfo
+                );
+            }
+            else
+            {
+                $this->template = ee()->extensions->call(
+                    'template_post_parse',
+                    $this->template,
+                    ($is_embed || $is_layout), // $is_partial
+                    $site_id,
+                    $currentTemplateInfo
```
The code seems to be identical, just assigned to different variables. Should it be moved to the top and only the variable assigment included inside conditional?

jcogs-design replied:
Layout was simply for clarity, for sure it can be compacted.

intoeetive commented on system/ee/legacy/libraries/Template.php:
```diff
@@ -232,15 +232,40 @@ public function fetch_and_parse($template_group = '', $template = '', $is_embed
+                );
+            }
+            // Add a conditional to adjust what is returned by function depending on what kind of template we are dealing with
+            if ($this->final_template)
```
I think this might work wrong if `$this->final_template` is an empty string, no?

intoeetive commented on system/ee/legacy/libraries/Template.php:
```diff
@@ -232,15 +232,40 @@ public function fetch_and_parse($template_group = '', $template = '', $is_embed
         }
         //
         // -------------------------------------------
+    return;
```
What is the reason this `return` has been added any why it does not return any value?

jcogs-design replied:
Simply to provide a point to add a debugging breakpoint.  Can be removed.

intoeetive reviewed (changes requested):
@jcogs-design I have added few comments on your code.

In addition, we are trying to follow PSR-12 rules, which require opening brace after `if` to be on same line.

jcogs-design committed:
Issue #1195 - Improved code
Hopefully this addresses issues flagged...
- PSR12 { position (corrected)
- if `ee('LivePreview')->hasEntryData()` missing (put back)
- code compacted somewhat
- conditional updated

jcogs-design commented:
Have posted updated code that I hope addresses your concerns.

jcogs-design committed:
Issue #1195 - fix typo
Final conditional was inverted - corrected.

intoeetive commented:
@jcogs-design looks like your branch (or repo) was deleted before we could merged your PR. Is there any chance you could restore it?

jcogs-design commented:
I think it is back now.

intoeetive commented:
![image](https://user-images.githubusercontent.com/752126/132644543-0715c265-4fc7-46d4-aa98-bb227aa2ecc3.png)
The tooltip says "this repository has been deleted"

jcogs-design commented:
No idea what is going on - so have created a new PR from the branch.

intoeetive commented:
Thank you!

Closing in favor of duplicate PR https://github.com/ExpressionEngine/ExpressionEngine/pull/1335
